collector/app.go

```diff
@@ -32,3 +105,64 @@ func GetAppsCollection(c *CollectorOpts, url string) *client.AppCollection {
 
 	return appCollection
 }
+
+func GetAllAppData(c *CollectorOpts, url string) []rancher.App {
+	appList := GetAppCollection(c, url)
+	if appList == nil {
+		return nil
+	}
+
+	appData := appList.Data
+	// Paginating if needed
+	appList, err := appList.Next()
+	for page := 2; appList != nil && err == nil; appList, err = appList.Next() {
+		log.Debug(" Paginating Applications page ", page)
+		appData = append(appData, appList.Data...)
+		page++
+	}
+	if err != nil {
+		log.Errorf(" Failed to paginate Applications err=%s", err)
+		return nil
+	}
+
+	return appData
+}
+
+func SplitAppExternalID(externalID string) map[string]string {
+	//Global catalog url: catalog://?catalog=demo&template=test&version=1.23.0
+	//Cluster catalog url: catalog://?catalog=c-XXXXX/test&type=clusterCatalog&template=test&version=1.23.0
+	//Project catalog url: catalog://?catalog=p-XXXXX/test&type=projectCatalog&template=test&version=1.23.0
+
+	str := strings.TrimPrefix(externalID, AppTemplateExternalIDPrefix)
+	values := strings.Split(str, "&")
+	out := make(map[string]string, len(values))
+	for _, v := range values {
+		pair := strings.Split(v, "=")
+		if len(pair) != 2 {
+			continue
+		}
+		if pair[0] == "catalog" && (strings.HasPrefix(pair[1], "c-") || strings.HasPrefix(pair[1], "p-")) {
+			pair[1] = strings.Replace(pair[1], "/", ":", -1)
+		}
+		out[pair[0]] = pair[1]
+	}
+
+	return out
+}
+
+func GetAppCatalogState(c *CollectorOpts, id string) string {
+	catalog, err := c.Client.Catalog.ByID(id)
+	if err != nil {
+		if IsNotFound(err) {
+			return "disabled"
+		}
+		log.Errorf("Failed to get Catalog ID %s err=%s", id, err)
+		return ""
```

Comment:
Is state for failed just empty string ? If you can't find out, might ask around for what it should be.

Reply:
Updated